README
QtSixA is the Sixaxis Joystick Manager
 ------------------------------------

QtSixA can connect PS3 hardware (Sixaxis/DualShock3 and Keypads)
to a Linux-compatible machine.

"QtSixA" is the GUI (interface), while "sixad" is the backend C code.

Currently sixad supports:
 - Sixaxis buttons, axis, accelerometers and LEDs
 - PS3 Keypads

sixad also registers the Sixaxis and Keypad's MAC/ID in the joystick name
(ex: "PLAYSTATION(R)3 Controller (00:XX:X0:0X:XX)".

sixad is triggered by udev, making it super easy to connect new devices
(you just need to press the PS button)

The Changelog is "debian/changelog" (the older one is "qtsixa/doc/changelog.old"

 -------------------------------------

QtSixA and sixad are licensed under the GNU GPL v2 license.

For help, check qtsixa/manual, or the online help at qtsixa.sourceforge.net


doc/changelog
qtsixa/sixad (1.0.1-falktx1) stable; urgency=low

  * Final release! (v1.0.1 for happy update)
  * QtSixA:
    - Final fix for the notifications
    - Fixed sixad-raw PATH (needed for RPM Linuxes)
    - Write to sixad config file as root (needed for RPM Linuxes)
    - Other very small fixes where needed
  * sixad:
    - hidraw-dump and sixad-raw do now read file instead of stdin
    [No more 'sudo su root -c "hid...', just 'hidraw-dump /dev/hidrawX']

 -- Filipe Coelho (aka "falkTX") <falktx@gmail.com>  29 Oct 2009 12:21:10 +0000


qtsixa/sixad (1.0.0-falktx7) stable; urgency=low

  * Pre-Final release (just for users to test)
  * QtSixA:
    - Do not apply a game profile if the it's configuration file doesn't exist
  * sixad:
    - Disable debug by default
    - Workaround for a PowerPC synaptics bug (LP#351059)

 -- Filipe Coelho (aka "falkTX") <falktx@gmail.com>  29 Oct 2009 13:02:17 +0000


qtsixa (1.0-falktx6) stable; urgency=low

  * Fixed the notifications not being able to start

-- Filipe Coelho (aka "falkTX") <falktx@gmail.com>  28 Oct 2009 09:21:24 +0000


qtsixa (1.0-falktx5) stable; urgency=low

  * QtSixA:
    - Fixed sixa-notify file appearing on home folder
    - Fixed notifications not auto-started on Gnome sessions
    - Notifications can now be started anytime by running "sixa-notify"
    - Disabled [Incomplete] Rumble feature (for v1.0 stability, can be enabled manually)
    - Display warning when "Legacy Driver" is not enabled on PowerPC

-- Filipe Coelho (aka "falkTX") <falktx@gmail.com>  26 Oct 2009 09:52:16 +0000


qtsixa/sixad (1.0-falktx4) stable; urgency=low

  * QtSixA:
    - Changed rumble feature to "Incomplete"
    - Added debug on/off option in "Configure Sixaxis"/"Advanced"
  * sixad:
    - *REALLY* Only process packets if received from a Sixaxis (Random uinput bug now fixed!)
    [print "got non-Sixaxis packet and ignored" for debug purposes]
    - Added debug on/off option

 -- Filipe Coelho (aka "falkTX") <falktx@gmail.com>  25 Oct 2009 20:22:57 +0000


qtsixa/sixad (1.0-falktx3) stable; urgency=low

  * QtSixA:
    - All bluetooth devices appear in the list, so updated features to match that
  * sixad:
    - Re-enabled axis deadzones
    - Deadzones also for Accelerometers and Accel/Speed/Position
    - The accelerometers are now properly centered
    - Small rumble effect along with LEDs animation
    - Fixed rumble activating all LEDs
    - Removed most rumble code for now (looking for the best way to implement it)
    - Only process packets if received from a Sixaxis (may fix random uinput bug)

 -- Filipe Coelho (aka "falkTX") <falktx@gmail.com>  25 Oct 2009 11:53:10 +0000


qtsixa (1.0-falktx2) stable; urgency=low

  * Removed "Status" from main device properties
  * Fixed "list of devices" unselectable when only USB devices were connected
  * Marked update as RC

 -- Filipe Coelho (aka "falkTX") <falktx@gmail.com>  23 Oct 2009 21:35:04 +0000


qtsixa (1.0-falktx1) stable; urgency=low

  * Auto-add *.fdi extension to new created profiles
  * Revert "Allow any filetype to be selected as fdi profile to add"
  * Fixed not showing fdi/png files in browser when adding a new profile (only happened on Gnome)

sixad (1.0-falktx{1,2}) stable; urgency=low

  * Fixed sixad search for devices
  * Trying to fix random input bug (disabling axis deadzones)

 -- Filipe Coelho (aka "falkTX") <falktx@gmail.com>  22 Oct 2009 17:31:26 +0000


qtsixa/sixad (1.0-falktx0) stable; urgency=low

  * QtSixA:
    - Changed homepage to 'http://qtsixa.sourceforge.net'
    - Fixed crash when creating a new profile using old PyQt (create button is now available)
    - Do not display warning about old PyQt anymore (bug fixed)
    - Fixed script to add new profile (it works now)
    - Allow any filetype to be selected as fdi profile to add
    - Finally wrote the manual!
    - A last check to the code, fixing all the possible bugs still around
  * sixad:
    - Small cleanup of code
    - Read sixad configuration file when connecting a new Sixaxis
    [You don't have to stop sixad to apply configuration changes]

 -- Filipe Coelho (aka "falkTX") <falktx@gmail.com>  21 Oct 2009 21:35:04 +0000


qtsixa (0.6.0-falktx2) stable; urgency=low

  * Fixed missing tray icon

sixad (0.9-falktx2) unstable; urgency=low

  * Reverse acX axis (was like this before, thus it makes more sense this way)

 -- Filipe Coelho (aka "falkTX") <falktx@gmail.com>  20 Oct 2009 13:58:27 +0000


qtsixa (0.6.0-falktx1) stable; urgency=low

  * Fixed bluetooth Keypad always appearing as USB Keypad, in notifications
  * Fixed some game profiles not applied properly
  * Updated "Configure Sixaxis" and "Sixaxis Reference" dialogs to match sixad updates
  * Added "Final Fantasy VIII PC" profile
  * Other small fixes as usual

sixad (0.9-falktx1) unstable; urgency=low

  * Preparations for v1.0 stable release: (remove unnecessary stuff, clean the rest)
    - Removed accelerometer buttons (not really needed and a little confusing)
    - Moved "gyro" to "accelerometers" section (but disabled for now)
    - Moved rumble (experimental) to "Input events"
    - Changed all "PS3 Fix" text to "Legacy driver"
    - Added "Position" axis (25-28)
    - Increased timeout for Acceleration/Speed/Position values
    - Cleaned up the accelerometers code (cause they can be disabled at any time)
  * Start 0,1,2,3 axis at 0
  * New option - set LED # according to joystick # (default now; js0 will be LED 1)
  * New option - start sixad at boot (disables normal bluetooth; can be stopped anytime)
  * Show sixad configuration when starting from terminal
  * Applied same updates for sixad-raw
  * Temporarily enable some debug output for 'sixad-uniput-sixaxis'
  * Fixed weird random input to axis 0-3
  * Disable Acceleration/Speed/Position by default (but keep Accelerometers)

 -- Filipe Coelho (aka "falkTX") <falktx@gmail.com>  20 Oct 2009 09:35:11 +0000


qtsixa (0.5.4-falktx1) stable; urgency=low

  * In the "Configure Sixaxis" dialog:

sixad-raw.c
/*
 * sixad-raw (Well, actually the same thing as sixad-uinput-sixaxis, but for hidraw devices instead of bluetooth)
 * written by falkTX, 2009  (used some code from xsixhidtest.c)
 *
 * Compile with: gcc -Wall sixad-raw.c -o sixad-raw
 * binary must be installed to /usr/sbin; then run with: "sixad --raw /dev/hidrawX"
 *
 *  TODO - read file instead of stdin
*/

#include <err.h>
#include <errno.h>
#include <fcntl.h>
#include <math.h>
#include <signal.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/ioctl.h>
#include <sys/stat.h>
#include <sys/time.h>
#include <sys/types.h>
#include <unistd.h>

#include <linux/input.h>
#include <linux/joystick.h>
#include <linux/uinput.h>

int enable_buttons;
int enable_sbuttons;
int enable_axis;
int enable_accel;
int enable_accon;
int enable_speed;
int enable_pos;
int ufd;

int just_started = 1;

struct device_info {
	unsigned int vendor;
	unsigned int product;
	unsigned int version;
};

struct state {
  double time;
  int ax, ay, az;       // Raw accelerometer data
  double ddx, ddy, ddz; // Acceleration
  double dx, dy, dz;    // Speed
  double x, y, z;       // Position
};

struct state prev;
struct state newH;
struct timeval tv;

/* UInput */
static char *uinput_filename[] = {"/dev/uinput", "/dev/input/uinput",
                           "/dev/misc/uinput"};
#define UINPUT_FILENAME_COUNT (sizeof(uinput_filename)/sizeof(char *))

int uinput_open(int enable_axis, int enable_accel)
{
    unsigned int i;
    int fd = -1;
    
    struct uinput_user_dev dev;
    
    /* Open uinput device */
    for (i=0; i < UINPUT_FILENAME_COUNT; i++) {
	if ((fd = open(uinput_filename[i], O_RDWR)) >= 0) {
	    break;
	}
    }
    
    ufd = fd;
    
    if (fd < 0) {
	fprintf(stderr, "Unable to open uinput\n");
	return -1;
    }
    
    memset(&dev, 0, sizeof(dev));
    
    dev.id.vendor = 0x054c;
    dev.id.product = 0x0268;
    dev.id.version = 0x0100;
    dev.id.bustype = BUS_USB;
		
    strncpy (dev.name, "PLAYSTATION(R)3 Controller (hidraw)", UINPUT_MAX_NAME_SIZE);

    ioctl (fd, UI_SET_EVBIT, EV_SYN);
    ioctl (fd, UI_SET_EVBIT, EV_ABS);
    ioctl (fd, UI_SET_EVBIT, EV_KEY);
    
    for (i = 0; i < 29; i++) {
        if (i >= 0 && i <= 3 && enable_axis == 1) { //left & right stick axis
            dev.absmax[i] = 127;
            dev.absmin[i] = -127;
        } else if (i == 4 && enable_accel == 1) { //accelerometer X
            dev.absmax[i] = 630;
            dev.absmin[i] = 410;
        } else if (i == 5 && enable_accel == 1) { //accelerometer Y
            dev.absmax[i] = 620;
            dev.absmin[i] = 400;
        } else if (i == 6 && enable_accel == 1) { //accelerometer Z
	  dev.absmax[i] = 630;
	  dev.absmin[i] = 418;
//      } else if (i == 7 && enable_accel == 1) { //Gyro
// 	dev.absmax[i] = ???;
// 	dev.absmin[i] = ???;
	} else if (i >= 8 && i <= 19) { //buttons
            dev.absmax[i] = 255;
            dev.absmin[i] = -255;
        } else if (i >= 20 && i <= 28) { //acceleration/speed/position
	    dev.absmax[i] = 1250;
	    dev.absmin[i] = -1250;
	} else {
            dev.absmax[i] = 32767;
            dev.absmin[i] = -32767;
	}

	if (ioctl (fd, UI_SET_ABSBIT, i) < 0) {
	    fprintf(stderr, "error on uinput ioctl (UI_SET_ABSBIT)\n");
	    return -1;
	}
    }

    for (i = 0; i < 17; i++) { 
	if (ioctl(fd, UI_SET_KEYBIT, BTN_JOYSTICK + i) < 0) {
	    fprintf(stderr, "error on uinput ioctl (UI_SET_KEYBIT)\n");
	    return -1;
	}
    }
	
    
    if (write(fd, &dev, sizeof(dev)) != sizeof(dev)) {
	fprintf(stderr, "Error on uinput device setup\n");
	close(fd);
	return -1;
    }

    if (ioctl(fd, UI_DEV_CREATE) < 0) {
	fprintf(stderr, "Error on uinput dev create");
	close(fd);
	return -1;
    }

    return fd;

}

int uinput_close(int fd)
{
    if (ioctl(fd, UI_DEV_DESTROY) < 0) {
	fprintf(stderr, "Error on uinput ioctl (UI_DEV_DESTROY)\n");
    }

    if (close(fd)) {
	fprintf(stderr, "Error on uinput close");
	return -1;
    }

    return 0;
}

int uinput_send(int fd, __u16 type, __u16 code, __s32 value)
{
    struct input_event event;

    memset(&event, 0, sizeof(event));
    event.type = type;
    event.code = code;
    event.value = value;
    gettimeofday(&(event.time), NULL);

    if (write(fd, &event, sizeof(event)) != sizeof(event)) {
	fprintf(stderr, "Error on send_event\n");
	return -1;
    }

    return 0;
}


void fatal(const char *msg) { perror(msg); exit(1); }

int main(int argc, char *argv[]) {
  
  struct state prev;
  memset(&prev, 0, sizeof(prev));
  
  unsigned char buf[128];
  int nr;
  int msg = 0;
  
  if (argc > 1) { //test feature
    if (atoi(argv[1]) == 3654) { //hidden codename for testing
      nr=read(0, buf, sizeof(buf));
      if ( nr < 0 ) { printf("Error: Not a device\n"); exit(1); }
      if ( nr == 50 ) { printf("Found a Sixaxis\n"); exit(1); }
      else { printf("Error: Not a Sixaxis\n"); exit(1); }
      exit(0);
    }
  }
  
  if (argc < 8) {
    printf("usage: sixad --raw <hidraw device>\n");
    exit(0);
  }
  
  enable_buttons = 1;
  enable_sbuttons = 1;
  enable_axis = 1;
  enable_accel = 1;
  enable_accon = 1;
  enable_speed = 1;
  enable_pos = 0;
  
  if (argc > 2) {
    enable_buttons = atoi(argv[2]);
    if (argc > 3) {
      enable_sbuttons = atoi(argv[3]);
      if (argc > 4) {
	enable_axis = atoi(argv[4]);
	if (argc > 5) 	{
	  enable_accel = atoi(argv[5]);
	  if (argc > 6) {
	    enable_accon = atoi(argv[6]);
	    if (argc > 7) {
	      enable_speed = atoi(argv[7]);
	      if (argc > 8)
		enable_pos = atoi(argv[8]);
	    }
	  }
	}
      }
    }
  }
  
  uinput_open(enable_axis, enable_accel);
  
  while ( (nr=read(0, buf, sizeof(buf))) ) {
    if ( nr < 0 ) { perror("Error when opening file"); exit(1); }
    if ( nr < 50 || nr > 52 ) { printf("Error: Not a Sixaxis!\n"); exit(1); }
    
    if ( msg == 0 ) { printf("Sixaxis sucessfully initiated\n"); msg = 1; }
    
    if ( gettimeofday(&tv, NULL) ) fatal("gettimeofday");
        newH.time = tv.tv_sec + tv.tv_usec*1e-6;
        newH.ax = buf[42]<<8 | buf[43];
        newH.ay = buf[44]<<8 | buf[45];
        newH.az = buf[46]<<8 | buf[47];
        if ( ! prev.time ) {
            prev.time = newH.time;
            prev.ax = newH.ax;
            prev.ay = newH.ay;
            prev.az = newH.az;
        }
        double dt = newH.time - prev.time; //(constants were recuced by half)
        double rc_dd = 1.0;  // Time constant for highpass filter on acceleration
        double alpha_dd = rc_dd / (rc_dd+dt);
        newH.ddx = alpha_dd*(prev.ddx + (newH.ax-prev.ax)*0.01);
        newH.ddy = alpha_dd*(prev.ddy + (newH.ay-prev.ay)*0.01);
        newH.ddz = alpha_dd*(prev.ddz - (newH.az-prev.az)*0.01);
        double rc_d = 1.0;  // Time constant for highpass filter on speed
        double alpha_d = rc_d / (rc_d+dt);
        newH.dx = alpha_d*(prev.dx + newH.ddx*dt);
        newH.dy = alpha_d*(prev.dy + newH.ddy*dt);
        newH.dz = alpha_d*(prev.dz + newH.ddz*dt);
        double rc = 0.5;  // Time constant for highpass filter on position
        double alpha = rc / (rc+dt);
        newH.x = alpha*(prev.x + newH.dx*dt);
        newH.y = alpha*(prev.y + newH.dy*dt);
        newH.z = alpha*(prev.z + newH.dz*dt);
        prev = newH;

        int b1 = buf[3];
        int b2 = buf[4];
        int b3 = buf[5];
        int lx = buf[7] - 128;
        int ly = buf[8] - 128;
        int rx = buf[9] - 128;
        int ry = buf[10] - 128;
        int acx = buf[42]<<8 | buf[43];
        int acy = buf[44]<<8 | buf[45];
        int acz = buf[46]<<8 | buf[47];
//     int gyro = buf[48]<<8 | buf[49];
        int up = buf[15];
        int right = buf[16];
        int down = buf[17];
        int left = buf[18];
        int l2 = buf[19];
        int r2 = buf[20];
        int l1 = buf[21];
        int r1 = buf[22];
        int tri = buf[23];
        int cir = buf[24];
        int cro = buf[25];
        int squ = buf[26];
        int posX = (int)(newH.x*1000);
        int posY = (int)(newH.y*1000);
        int posZ = (int)(newH.z*1000);
        int accX = (int)(newH.ddx*1000);
        int accY = (int)(newH.ddy*1000);
        int accZ = (int)(newH.ddz*1000);
        int velX = (int)(newH.dx*1000);
        int velY = (int)(newH.dy*1000);
        int velZ = (int)(newH.dz*1000);

	//deadzones - NOTE - this may cause garbage input...!
	if ((lx > -10 && lx < 10) || lx > 255) { lx = 0; }
	if ((ly > -10 && ly < 10) || ly > 255) { ly = 0; }
	if ((rx > -10 && rx < 10) || rx > 255) { rx = 0; }
	if ((ry > -10 && ry < 10) || ry > 255) { ry = 0; }
//  	if (acx > 508 && acx < 516) { acx = 512; }
// 	if (acy > 508 && acy < 516) { acy = 512; }
// 	if (acz > 508 && acz < 516) { acz = 512; }
//  	if (posX > -150 && posX < -150) { posX = 0; }
// 	if (posY > -150 && posY < -150) { posY = 0; }
// 	if (posZ > -150 && posZ < -150) { posZ = 0; }
// 	if (accX > -150 && accX < -150) { accX = 0; }
// 	if (accY > -150 && accY < -150) { accY = 0; }
// 	if (accZ > -150 && accZ < -150) { accZ = 0; }
// 	if (velX > -150 && velX < -150) { velX = 0; }
// 	if (velY > -150 && velY < -150) { velY = 0; }
// 	if (velZ > -150 && velZ < -150) { velZ = 0; }

        // buttons
        if (enable_buttons == 1) {
            uinput_send(ufd, EV_KEY, BTN_JOYSTICK + 0, b1 & 1);
            uinput_send(ufd, EV_SYN, SYN_REPORT, 0);
            b1 >>= 1;
            uinput_send(ufd, EV_KEY, BTN_JOYSTICK + 1, b1 & 1);
            uinput_send(ufd, EV_SYN, SYN_REPORT, 0);
            b1 >>= 1;
            uinput_send(ufd, EV_KEY, BTN_JOYSTICK + 2, b1 & 1);
            uinput_send(ufd, EV_SYN, SYN_REPORT, 0);
            b1 >>= 1;
            uinput_send(ufd, EV_KEY, BTN_JOYSTICK + 3, b1 & 1);
            uinput_send(ufd, EV_SYN, SYN_REPORT, 0);
            b1 >>= 1;
            uinput_send(ufd, EV_KEY, BTN_JOYSTICK + 4, b1 & 1);
            uinput_send(ufd, EV_SYN, SYN_REPORT, 0);
            b1 >>= 1;
            uinput_send(ufd, EV_KEY, BTN_JOYSTICK + 5, b1 & 1);
            uinput_send(ufd, EV_SYN, SYN_REPORT, 0);
            b1 >>= 1;
            uinput_send(ufd, EV_KEY, BTN_JOYSTICK + 6, b1 & 1);
            uinput_send(ufd, EV_SYN, SYN_REPORT, 0);
            b1 >>= 1;
            uinput_send(ufd, EV_KEY, BTN_JOYSTICK + 7, b1 & 1);
            uinput_send(ufd, EV_SYN, SYN_REPORT, 0);
            uinput_send(ufd, EV_KEY, BTN_JOYSTICK + 8, b2 & 1);
            uinput_send(ufd, EV_SYN, SYN_REPORT, 0);
            b2 >>= 1;
            uinput_send(ufd, EV_KEY, BTN_JOYSTICK + 9, b2 & 1);
            uinput_send(ufd, EV_SYN, SYN_REPORT, 0);
            b2 >>= 1;
            uinput_send(ufd, EV_KEY, BTN_JOYSTICK + 10, b2 & 1);
            uinput_send(ufd, EV_SYN, SYN_REPORT, 0);
            b2 >>= 1;
            uinput_send(ufd, EV_KEY, BTN_JOYSTICK + 11, b2 & 1);
            uinput_send(ufd, EV_SYN, SYN_REPORT, 0);
            b2 >>= 1;
            uinput_send(ufd, EV_KEY, BTN_JOYSTICK + 12, b2 & 1);
            uinput_send(ufd, EV_SYN, SYN_REPORT, 0);
            b2 >>= 1;
            uinput_send(ufd, EV_KEY, BTN_JOYSTICK + 13, b2 & 1);
            uinput_send(ufd, EV_SYN, SYN_REPORT, 0);
            b2 >>= 1;
            uinput_send(ufd, EV_KEY, BTN_JOYSTICK + 14, b2 & 1);
            uinput_send(ufd, EV_SYN, SYN_REPORT, 0);
            b2 >>= 1;
            uinput_send(ufd, EV_KEY, BTN_JOYSTICK + 15, b2 & 1);
            uinput_send(ufd, EV_SYN, SYN_REPORT, 0);
            uinput_send(ufd, EV_KEY, BTN_JOYSTICK + 16, b3 & 1);
            uinput_send(ufd, EV_SYN, SYN_REPORT, 0);
        }

        //axis
        if (enable_axis == 1) {
            uinput_send(ufd, EV_ABS, 0, lx);
            uinput_send(ufd, EV_SYN, SYN_REPORT, 0);
            uinput_send(ufd, EV_ABS, 1, ly);
            uinput_send(ufd, EV_SYN, SYN_REPORT, 0);
            uinput_send(ufd, EV_ABS, 2, rx);
            uinput_send(ufd, EV_SYN, SYN_REPORT, 0);
            uinput_send(ufd, EV_ABS, 3, ry);
            uinput_send(ufd, EV_SYN, SYN_REPORT, 0);
        }

        //accelerometer RAW
        if (enable_accel == 1) {
            uinput_send(ufd, EV_ABS, 4, acx);
            uinput_send(ufd, EV_SYN, SYN_REPORT, 0);
            uinput_send(ufd, EV_ABS, 5, acy);
            uinput_send(ufd, EV_SYN, SYN_REPORT, 0);
            uinput_send(ufd, EV_ABS, 6, acz);
            uinput_send(ufd, EV_SYN, SYN_REPORT, 0);
//     uinput_send(ufd, EV_ABS, 7, gyro);
//     uinput_send(ufd, EV_SYN, SYN_REPORT, 0);
        }

        //buttons (sensible, as axis)
        if (enable_sbuttons == 1) {
            uinput_send(ufd, EV_ABS, 8, up);
            uinput_send(ufd, EV_SYN, SYN_REPORT, 0);
            uinput_send(ufd, EV_ABS, 9, right);
            uinput_send(ufd, EV_SYN, SYN_REPORT, 0);
            uinput_send(ufd, EV_ABS, 10, down);
            uinput_send(ufd, EV_SYN, SYN_REPORT, 0);
            uinput_send(ufd, EV_ABS, 11, left);
            uinput_send(ufd, EV_SYN, SYN_REPORT, 0);
            uinput_send(ufd, EV_ABS, 12, l2);
            uinput_send(ufd, EV_SYN, SYN_REPORT, 0);
            uinput_send(ufd, EV_ABS, 13, r2);
            uinput_send(ufd, EV_SYN, SYN_REPORT, 0);
            uinput_send(ufd, EV_ABS, 14, l1);
            uinput_send(ufd, EV_SYN, SYN_REPORT, 0);
            uinput_send(ufd, EV_ABS, 15, r1);
            uinput_send(ufd, EV_SYN, SYN_REPORT, 0);
            uinput_send(ufd, EV_ABS, 16, tri);
            uinput_send(ufd, EV_SYN, SYN_REPORT, 0);
            uinput_send(ufd, EV_ABS, 17, cir);
            uinput_send(ufd, EV_SYN, SYN_REPORT, 0);
            uinput_send(ufd, EV_ABS, 18, cro);
            uinput_send(ufd, EV_SYN, SYN_REPORT, 0);
            uinput_send(ufd, EV_ABS, 19, squ);
            uinput_send(ufd, EV_SYN, SYN_REPORT, 0);
        }

        //acceleration
        if (enable_accon == 1) {
            uinput_send(ufd, EV_ABS, 20, accX);
            uinput_send(ufd, EV_SYN, SYN_REPORT, 0);
            uinput_send(ufd, EV_ABS, 21, accY);
            uinput_send(ufd, EV_SYN, SYN_REPORT, 0);
            uinput_send(ufd, EV_ABS, 22, accZ);
            uinput_send(ufd, EV_SYN, SYN_REPORT, 0);
        }

        //speed
        if (enable_speed == 1) {
            uinput_send(ufd, EV_ABS, 23, velX);
            uinput_send(ufd, EV_SYN, SYN_REPORT, 0);
            uinput_send(ufd, EV_ABS, 24, velY);
            uinput_send(ufd, EV_SYN, SYN_REPORT, 0);
            uinput_send(ufd, EV_ABS, 25, velZ);
            uinput_send(ufd, EV_SYN, SYN_REPORT, 0);
        }

        //position
        if (enable_pos == 1) {
            uinput_send(ufd, EV_ABS, 26, posX);
            uinput_send(ufd, EV_SYN, SYN_REPORT, 0);
            uinput_send(ufd, EV_ABS, 27, posY);
            uinput_send(ufd, EV_SYN, SYN_REPORT, 0);
            uinput_send(ufd, EV_ABS, 28, posZ);
            uinput_send(ufd, EV_SYN, SYN_REPORT, 0);
        }

  }
  return 0;
} //The End
